### PR TITLE
Set up user API

### PR DESCRIPTION
**Pull request by NganJason**
This pull request involves
- User routers
- User model and methods
- User route handler
   - login handler
   - signup handler
   - forgot password handler
   - reset password handler
- custom ErrorResponse middleware

### CODE REVIEW

File: backend/middleware/auth.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  let token;
+
+  if (
+    req.headers.authorization &&
+    req.headers.authorization.startsWith("Bearer")
+  ) {
+    token = req.headers.authorization.split(" ")[1];
+  }
+
+  if (!token) {
```
Comment (bangyii) [inline backend/middleware/auth.js:15]: Can straight try verify token below? Wrong token will throw error kan

Comment (NganJason) [reply backend/middleware/auth.js:15]: You are right, let me remove it

File: backend/server.js [hunk 2]
```diff
@@ -1,26 +1,41 @@
...
-    await mongoose.connect(mongoUri, { useNewUrlParser: true, useUnifiedTopology: true })
-    console.log('Successfully connected to database!')
+    await mongoose.connect(mongoUri, {
+      useNewUrlParser: true,
+      useUnifiedTopology: true,
+    });
+    console.log("Successfully connected to database!");
+
+    // For parsing cookies
+    app.use(cookieParser());
```
Comment (bangyii) [inline backend/server.js:23]: This one can put inside app.js?

Comment (NganJason) [reply backend/server.js:23]: okay

File: backend/handlers/userHandler.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+export const resetPassword = async (req, res, next) => {
+  const signedToken = req.params.signedToken;
+
+  try {
+    const decoded = jwt.verify(signedToken, process.env.JWT_SECRET);
+    var d = new Date();
+
+    if (decoded.exp < d.getTime() / 1000) {
+      return next(new ErrorResponse("Token expired", 400));
```
Comment (bangyii) [inline backend/handlers/userHandler.js:98]: If the signed token has expiresIn, then decode should automatically throw error if the token you're trying to verify is expired, no need manual check

File: backend/middleware/auth.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+import ErrorResponse from "../utils/errorResponse.js";
+
+export const isAuthenticated = async (req, res, next) => {
+  const token = req.cookies.token;
+
+  try {
+    const decoded = jwt.verify(token, process.env.JWT_SECRET);
+    var d = new Date();
+
+    if (decoded.exp < d.getTime() / 1000) {
```
Comment (bangyii) [inline backend/middleware/auth.js:12]: This one also, if expiresIn is defined, no need to check date

File: backend/handlers/userHandler.js (hunk 3 above)
Comment (NganJason) [reply backend/handlers/userHandler.js]: You are right, I removed it

**Pull request merged**